ManagedHooksManager/ApiHook.cs

```diff
@@ -0,0 +1,198 @@
+
+                var structSize = NativeApi.VirtualQuery(
+                    memoryBlockPtr,
+                    memoryInfoStructPtr,
+                    new IntPtr(memoryInfoStructSize));
+
+                if (structSize == IntPtr.Zero)
+                    throw Marshal.GetExceptionForHR(Marshal.GetHRForLastWin32Error());
+
+                var memoryInfo = Marshal.PtrToStructure<
+                    NativeStructs.MEMORY_BASIC_INFORMATION64>(memoryInfoStructPtr);
+
+                if (memoryInfo.State == NativeConstants.MEM_FREE &&
+                    memoryInfo.RegionSize >= trampoline.Length)
+                {
+                    trampolineAddress = NativeApi.VirtualAlloc(
+                        memoryBlockPtr,
+                        trampolineSize,
+                        NativeConstants.MEM_COMMIT | NativeConstants.MEM_RESERVE,
+                        NativeConstants.PAGE_READWRITE);
+                }
+            }
+            while (trampolineAddress == IntPtr.Zero &&
+                   tryMemoryBlock - moduleBaseAddress <= int.MaxValue);
+
+            Marshal.FreeHGlobal(memoryInfoStructPtr);
+
+            if (trampolineAddress == IntPtr.Zero)
+                throw new InvalidOperationException("Failed to allocate space for trampoline in +/-2GB range.");
+
+            Marshal.Copy(trampoline, 0, trampolineAddress, trampoline.Length);
+            int newFuncRva = Convert.ToInt32(trampolineAddress.ToInt64() - moduleBaseAddress);
+
+            var protectResult = NativeApi.VirtualProtect(
+                trampolineAddress,
+                trampolineSize,
+                NativeConstants.PAGE_EXECUTE_READ,
+                out _);
+
+            if (!protectResult)
+                throw Marshal.GetExceptionForHR(Marshal.GetHRForLastWin32Error());
+
+            return newFuncRva;
+        }
+
+        private static byte[] GetTrampolineInstructions(IntPtr newFuncPointer)
+        {
+            var processorArchitecture = Environment.GetEnvironmentVariable("PROCESSOR_ARCHITECTURE");
+
+            if (string.Equals(processorArchitecture, "amd64", StringComparison.OrdinalIgnoreCase))
+            {
+                var trampoline = new byte[]
+                {
+                    0x68, 0xcc, 0xcc, 0xcc, 0xcc,                       // push   dword low
+                    0xc7, 0x44, 0x24, 0x04, 0xcc, 0xcc, 0xcc, 0xcc,     // mov    DWORD PTR [rsp+4], dword high
+                    0xc3                                                // ret
+                };
+
+                var newFuncAddressBytes = BitConverter.GetBytes(newFuncPointer.ToInt64());
+
+                Array.Copy(newFuncAddressBytes, 0, trampoline, 1, 4);
+                Array.Copy(newFuncAddressBytes, 4, trampoline, 9, 4);
+
+                return trampoline;
+            }
+
+            throw new PlatformNotSupportedException($"Processor architecture {processorArchitecture} is not supported.");
```

Comment:
There is a code for ARM64 architecture, but I didn't commit it yet as there is no ARM64 .NET Framework currently.